### DISCUSSION TITLE
TuyaMCU mapping for heater

### DISCUSSION THREAD

**Discussion by Tweak-NL** [arendst/Tasmota#13547] [Ideas]
Hello, maybe stupid question.. I got a Eurom Alutherm heater from a friend because the Tuya app was losing connection all the time. 

And indeed, same problem here.. So i took it apart made a full backup and flashed Tasmota on it and wow it works and the losing connection problem seems gone. But the FunctionId's are not correct. TuyaMCU 7,102 does not work. I think the screenshots will show what the problem is:

![mapping-ids](https://user-images.githubusercontent.com/67078330/139955335-5df2b45a-5629-42b0-af40-b0da4b36d6d2.JPG)
![ids](https://user-images.githubusercontent.com/67078330/139955336-3655300b-cc5f-43dd-bbcd-971402faa05f.JPG)

(the heater looks like this one https://templates.blakadder.com/klarstein_10034390.html)

ESP inside:

![TYJW2](https://user-images.githubusercontent.com/67078330/139963108-96c2a9a2-d693-4e95-9fa2-1f4fa59c728e.jpeg)

Maybe this "protocol" is different then the others.

**Comment by Tweak-NL** [comment]
Why is this placed in idea's ? I think i made a mistake, maybe this is just not possible because this heater/mcu is not supported?

**Comment by blakadder** [comment]
Its evident your control board is different. dpid functions above 100 are manufacturer specific so you need to find their functions yourself by testing using appropriate tuyasend commands

With TuyaMCU 7,102 you're trying to map fnid7 which is not available to dpid 102, that's why it doesn't work

**Comment by Tweak-NL** [comment]
Hmm, so this can never work correctly with MQTT? I know some commands example:

curl http://192.168...../cm?cmnd=TuyaSend4%20101,02 <- this will set the heating mode to medium

101

00 = low
01 = medium
02 = high
03 = ??

And in domoticz i use: domoticz.openURL(string.format('http://192.168..../cm?cmnd=TuyaSend2+2,%s', device.setPoint))
This all works.. Now the problem.

If i send a new temperature to the device while its in off mode it will turn on but dpid **1** is not updated. So i dont really know it's really on? (yeah only when i check my power usage from the house)

When the heater changes from mode high (02) to medium (01) MQTT is not updated with this info i think? Turning the heater on and off works perfectly with MQTT

**Comment by Tweak-NL** [comment]
Here what happends when i send a new temp to the heater while it's in OFF mode:

15:18:16.469 SRC: WebCommand from 192.168.
15:18:16.472 CMD: Grp 0, Cmnd 'TUYASEND', Idx 2, Len 6, Data '2,17.0'
15:18:16.492 TYA: Send "55aa00060008020200040000001126"
15:18:16.497 MQT: stat/tasmota_12B94A/RESULT = {"TuyaSend":"Done"}
15:18:16.608 {"TuyaReceived":{"Data":"55AA01070008020200040000001128","Cmnd":7,"CmndData":"0202000400000011","DpType2Id2":17,"2":{"DpId":2,"DpIdType":2,"DpIdData":"00000011"}}}
15:18:16.612 TYA: fnId=72 is set for dpId=2
15:18:16.617 MQT: tele/tasmota_12B94A/SENSOR = {"TuyaSNS":{"TempSet":17}}
15:18:16.619 TYA: RX value 17 from dpId 2 
15:18:16.625 {"TuyaReceived":{"Data":"55AA01070008030200040000000A22","Cmnd":7,"CmndData":"030200040000000A","DpType2Id3":10,"3":{"DpId":3,"DpIdType":2,"DpIdData":"0000000A"}}}
15:18:16.628 TYA: fnId=71 is set for dpId=3
15:18:16.630 TYA: RX value 10 from dpId 3 
15:18:16.636 {"TuyaReceived":{"Data":"55AA01070005040400010116","Cmnd":7,"CmndData":"0404000101","DpType4Id4":1,"4":{"DpId":4,"DpIdType":4,"DpIdData":"01"}}}
15:18:16.639 TYA: fnId=0 is set for dpId=4
15:18:16.648 {"TuyaReceived":{"Data":"55AA01070005650400010278","Cmnd":7,"CmndData":"6504000102","DpType4Id101":2,"101":{"DpId":101,"DpIdType":4,"DpIdData":"02"}}}
15:18:16.652 TYA: fnId=1 is set for dpId=101
15:18:16.662 {"TuyaReceived":{"Data":"55AA01070005660100010074","Cmnd":7,"CmndData":"6601000100","DpType1Id102":0,"102":{"DpId":102,"DpIdType":1,"DpIdData":"00"}}}
15:18:16.665 TYA: fnId=12 is set for dpId=102
15:18:16.667 TYA: RX Relay-2 --> MCU State: Off Current State:Off
15:18:16.673 {"TuyaReceived":{"Data":"55AA010700050C050001001E","Cmnd":7,"CmndData":"0C05000100","DpType5Id12":"0x00","12":{"DpId":12,"DpIdType":5,"DpIdData":"00"}}}
15:18:16.676 TYA: fnId=0 is set for dpId=12
15:18:17.610 {"TuyaReceived":{"Data":"55AA01070008020200040000001128","Cmnd":7,"CmndData":"0202000400000011","DpType2Id2":17,"2":{"DpId":2,"DpIdType":2,"DpIdData":"00000011"}}}
15:18:17.614 TYA: fnId=72 is set for dpId=2
15:18:17.616 TYA: RX value 17 from dpId 2 
15:18:17.626 {"TuyaReceived":{"Data":"55AA01070008030200040000000A22","Cmnd":7,"CmndData":"030200040000000A","DpType2Id3":10,"3":{"DpId":3,"DpIdType":2,"DpIdData":"0000000A"}}}
15:18:17.630 TYA: fnId=71 is set for dpId=3
15:18:17.632 TYA: RX value 10 from dpId 3 
15:18:17.638 {"TuyaReceived":{"Data":"55AA01070005040400010116","Cmnd":7,"CmndData":"0404000101","DpType4Id4":1,"4":{"DpId":4,"DpIdType":4,"DpIdData":"01"}}}
15:18:17.641 TYA: fnId=0 is set for dpId=4
15:18:17.650 {"TuyaReceived":{"Data":"55AA01070005650400010278","Cmnd":7,"CmndData":"6504000102","DpType4Id101":2,"101":{"DpId":101,"DpIdType":4,"DpIdData":"02"}}}
15:18:17.653 TYA: fnId=1 is set for dpId=101
15:18:17.663 {"TuyaReceived":{"Data":"55AA01070005660100010074","Cmnd":7,"CmndData":"6601000100","DpType1Id102":0,"102":{"DpId":102,"DpIdType":1,"DpIdData":"00"}}}
15:18:17.666 TYA: fnId=12 is set for dpId=102
15:18:17.668 TYA: RX Relay-2 --> MCU State: Off Current State:Off
15:18:17.675 {"TuyaReceived":{"Data":"55AA010700050C050001001E","Cmnd":7,"CmndData":"0C05000100","DpType5Id12":"0x00","12":{"DpId":12,"DpIdType":5,"DpIdData":"00"}}}
15:18:17.679 TYA: fnId=0 is set for dpId=12
15:18:18.682 WIF: Checking connection...
15:18:20.739 DOM: idx 298, nvalue 0
15:18:20.839 DOM: idx 299, nvalue 0
15:18:20.888 DOM: idx 300, nvalue 0
15:18:20.938 DOM: idx 302, nvalue 0
15:18:20.989 DOM: idx 304, nvalue 0
15:18:21.039 DOM: idx 305, nvalue 0
15:18:21.088 DOM: idx 307, nvalue 0
15:18:21.137 DOM: idx 308, nvalue 0
15:18:21.188 DOM: idx 309, nvalue 0
15:18:21.237 DOM: idx 310, nvalue 0
15:18:21.392 TYA: Send "55aa00000000ff"
15:18:25.931 DOM: idx 298, nvalue 0

Here u can see dpid 1 is not getting updated/missing.. (maybe that is really something broken on this heater?)
![missing1](https://user-images.githubusercontent.com/67078330/141130178-d388c577-ea00-44ca-8481-0b3c3b5ff06e.JPG)

**Comment by blakadder** [comment]
if its not updated the tuyamcu doesn't want to update it.

you can try sending tuyasend0 or 8 periodically or on certain commands and see if that polls dpid successfully 

every command that works over http will work using mqtt

**Reply by Tweak-NL** [reply]
Thnx wow that seems to work! i turned it on with sending a temp and tried tuyasend0 in console.
And indeed now dpid 1 is updated and my Domoticz is saying the heater is on. Is there a timer build in i can use?

And that about MQTT sounds good but ill show u i cannot "enable" it and define a IDX for domoticz. Because the DPID is such a high number. 
![idx](https://user-images.githubusercontent.com/67078330/141132641-efad66cc-41cc-4c88-95d1-86716dd78546.JPG)

**Reply by blakadder** [reply]
using domoticz != mqtt

you can create a rule that triggers on whatever message domoticz sends over mqtt

**Reply by Tweak-NL** [reply]
Hmm, it seems timers are build in to? 

But this: Rule1 ON Time#Minute DO tuyasend0 ENDON does not seem to work.. 
And yes i know domoticz is not mqtt but it should work perfectly like my zigbee devices.
And it does work with the DPID 1 @ IDX 336 (domoticz switch)
